feedback/invalid_token.go

```diff
@@ -0,0 +1,224 @@
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package feedback
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	raven "github.com/getsentry/raven-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"github.com/topfreegames/pusher/extensions"
+	"github.com/topfreegames/pusher/interfaces"
+	"github.com/topfreegames/pusher/util"
+)
+
+// InvalidToken represents a token with the necessary information to be deleted
+type InvalidToken struct {
+	Token    string
+	Game     string
+	Platform string
+}
+
+// InvalidTokenHandler takes the InvalidTokens from the InChannel and put them in a buffer.
+// When the buffer is full or after a timeout, it is flushed, triggering the deletion
+// of the tokens from the database
+type InvalidTokenHandler struct {
+	Logger *log.Logger
+	Config *viper.Viper
+	Client *extensions.PGClient
+
+	FlushTicker *time.Ticker
+	InChan      *chan *InvalidToken
+	Buffer      []*InvalidToken
+	bufferSize  int
+	run         bool
+	stopChan    chan bool
+}
+
+// NewInvalidTokenHandler returns a new InvalidTokenHandler instance
+func NewInvalidTokenHandler(
+	logger *log.Logger, cfg *viper.Viper,
+	inChan *chan *InvalidToken,
+	dbOrNil ...interfaces.DB,
+
+) (*InvalidTokenHandler, error) {
+	h := &InvalidTokenHandler{
+		Logger:   logger,
+		Config:   cfg,
+		InChan:   inChan,
+		stopChan: make(chan bool),
+	}
+
+	var db interfaces.DB
+	if len(dbOrNil) == 1 {
+		db = dbOrNil[0]
+	}
+
+	err := h.configure(db)
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+func (i *InvalidTokenHandler) loadConfigurationDefaults() {
+	i.Config.SetDefault("feedbackListeners.invalidToken.flush.time.ms", 5000)
+	i.Config.SetDefault("feedbackListeners.invalidToken.buffer.size", 1000)
+}
+
+func (i *InvalidTokenHandler) configure(db interfaces.DB) error {
+	l := i.Logger.WithField("method", "configure")
+	i.loadConfigurationDefaults()
+
+	flushTime := time.Duration(i.Config.GetInt("feedbackListeners.invalidToken.flush.time.ms")) * time.Millisecond
+	i.bufferSize = i.Config.GetInt("feedbackListeners.invalidToken.buffer.size")
+
+	i.FlushTicker = time.NewTicker(flushTime)
```

Comment:
we must always defer ticker.Stop() when initializing a new ticker, otherwise we'll have CPU leaks

Reply:
I'll change that. Thanks!